### PR TITLE
cmake: zephyr modules: sanitize all module name when used as variable

### PR DESCRIPTION
**Pull request by tejlmand**
The introduction of Zephyr module glue code in the Zephyr repository
introduces a Kconfig variable in the form of:
`config ZEPHYR_<MODULE_NAME>_MODULE`.

All Kconfig variables go into `autoconf.h`, therefore it is necessary
to sanitize the Kconfig variable, so that it does not contain special
characters. To ensure consistent variable name, then the module name
will be sanitized in all variable use in both Kconfig and CMake.
The sanitization is done be replacing all special characters with an
underscore, `_`.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

### CODE REVIEW

File: cmake/extensions.cmake [hunk 1]
```diff
@@ -1960,6 +1960,47 @@ Relative paths are only allowed with `-D${ARGV1}=<path>`")
...
+# Usage:
+#   zephyr_string(<mode> <out-var> <input> ...)
+#
+# Zephyr string function extension.
+# This function extends the CMake string function by providing additional
+# manipulation arguments to CMake string.
+#
+# SANITIZE: Ensure that the output string does not contain any special
+#           characters. Special characters, such as -, +, =, $, etc. are
+#           converted to an underscores '_'.
```
Comment (anangl) [inline cmake/extensions.cmake:1972]:
```suggestion
#           converted to underscores '_'.
```
Same below.

File: doc/guides/modules.rst [hunk 2]
```diff
@@ -421,6 +421,12 @@ CMake variable ``ZEPHYR_<MODULE_NAME>_MODULE_DIR`` and the variable
 ``ZEPHYR_<MODULE_NAME>_CMAKE_DIR`` holds the location of the directory
 containing the module's :file:`CMakeLists.txt` file.
 
+.. note::
+   All symbols in module names will be replaced with an underscore, ``_``, and
+   converted to upper characters when used for CMake and Kconfig variables.
```
Comment (anangl) [inline doc/guides/modules.rst:426]: This sentence could be rephrased, as right now it is a bit unclear.

**Review by anangl — COMMENTED** [review]
Looks good. Some nits only spotted.

File: doc/guides/modules.rst (hunk 2 above)
Comment (tejlmand) [reply doc/guides/modules.rst:426]: How about:
>  The CMake and Kconfig variables will be the module name converted to upper characters, and all symbols will be replaced with an underscore, ``_``.

is that clearer ?

Comment (anangl) [reply doc/guides/modules.rst:426]: Maybe:
> When used for CMake and Kconfig variables, all letters in module names are
converted to uppercase and all non-alphanumeric characters are converted
to underscores (``_``).

File: cmake/extensions.cmake (hunk 1 above)
Comment (anangl) [reply cmake/extensions.cmake]: There's one more such "an" [right below](https://github.com/zephyrproject-rtos/zephyr/pull/31422/files#diff-c4994ce9c473874bf18eb76dc67b5cfc1ecc2171de4d4ddef83ddfdab6bef559R1976).

File: doc/guides/modules.rst (hunk 2 above)
Comment (tejlmand) [reply doc/guides/modules.rst:426]: done.

**Review by anangl — APPROVED** [review]

File: cmake/extensions.cmake (hunk 1 above)
Comment (tejlmand) [reply cmake/extensions.cmake]: and fixed.

File: cmake/extensions.cmake [hunk 3]
```diff
@@ -1960,6 +1960,47 @@ Relative paths are only allowed with `-D${ARGV1}=<path>`")
...
+# This function extends the CMake string function by providing additional
+# manipulation arguments to CMake string.
+#
+# SANITIZE: Ensure that the output string does not contain any special
+#           characters. Special characters, such as -, +, =, $, etc. are
+#           converted to underscores '_'.
+#
+# SANITIZE TOUPPER: Ensure that the output string does not contain any special
+#                   characters. Special characters, such as -, +, =, $, etc. are
+#                   converted to underscores '_'.
```
Comment (ioannisg) [inline cmake/extensions.cmake:1976]: @tejlmand what is the actual impact of this conversion? No objections, but would like to know if there are modules that actually have special characters in the names :)

**Review by ioannisg — APPROVED** [review]

File: cmake/extensions.cmake (hunk 3 above)
Comment (carlescufi) [reply cmake/extensions.cmake:1976]: Yes, we have modules with `-` in their names. But the path won't change with this PR, just the internal name used by CMake and Kconfig.

**Review by carlescufi — APPROVED** [review]

File: cmake/extensions.cmake (hunk 3 above)
Comment (tejlmand) [reply cmake/extensions.cmake:1976]: as we now introduces the module as Kconfig variable for proper glue code handling outside of the module, then we need to sanitize the name, as it goes into `autoconf`, like this:
```
.../build/zephyr/include/generated/autoconf.h:37:9: warning: ISO C99 requires whitespace after the macro name
   37 | #define CONFIG_ZEPHYR_TRUSTED-FIRMWARE-M_MODULE 1
```
we could of course just sanitize for Kconfig symbols, but I would like that all places are using same form, so therefore sanitizing everywhere.

**Pull request merged**